activerecord/lib/active_record/reflection.rb

```diff
@@ -1105,7 +1105,7 @@ def alias_candidate(name)
       end
 
       def alias_name
-        Arel::Table.new(table_name)
+        Arel::Table.new(table_name, type_caster: TypeCaster::Map.new(klass))
```

Comment:
Sorry, I thought `type_caster` was a private method like the other Arel type methods. Updated my commit